### PR TITLE
add dependency on typing_extensions, add ruff, update deps

### LINKED ISSUE #61: `T_in` and `T_out` `TypeVar`s
`in` / `out` is based

### LINKED ISSUE #64: python 3.12 support

### LINKED ISSUE #68: include an `__all__`

### LINKED ISSUE #73: pylint -> ruff (everything else as well)

### LINKED ISSUE #79: depend on `typing_extensions`

### PR DESCRIPTION
**Pull request by KotlinIsland**
- resolves #73
- resolves #61
- resolves #79
- resolves #64
- resolves #68

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -40,28 +34,10 @@ skip-magic-trailing-comma = true
...
-[tool.pylint.FORMAT]
-max-line-length = 200
-
 [tool.pytest.ini_options]
 addopts = "--import-mode=importlib"
 xfail_strict = true
+
+[tool.ruff]
+extend-select = ["I", "UP"]
+ignore = ["E501"]
```
Comment (DetachHead) [inline pyproject.toml]: maybe copy the ruff config from https://github.com/DetachHead/pytest-robotframework/blob/master/pyproject.toml

File: pyproject.toml [hunk 2]
```diff
@@ -40,28 +34,10 @@ skip-magic-trailing-comma = true
...
-    "duplicate-code",                            # https://github.com/PyCQA/pylint/issues/214
-    "isinstance-second-argument-not-valid-type", # false positive on Never type, handled by mypy
-    "no-value-for-parameter",                    # handled by mypy
-    "too-many-return-statements",
-]
-enable = ["useless-suppression", "deprecated-pragma"]
-
-[tool.pylint.FORMAT]
-max-line-length = 200
-
```
Comment (DetachHead) [inline pyproject.toml]: is it worth keeping pylint around until all the rules are in ruff? (probably not)

**Comment by DetachHead** [comment]
can u also add vscode ruff tasks plz

https://github.com/DetachHead/pytest-robotframework/blob/67d5dce519809ba7442ec95069c8afa152b612d4/.vscode/tasks.json#L55-L74

File: pyproject.toml (hunk 2 above)
Comment (KotlinIsland) [reply pyproject.toml]: not, good riddance

File: basedtyping/__init__.py [hunk 3]
```diff
@@ -58,15 +51,13 @@ def __getitem__(self, item):
     Function = Callable
 
 # Unlike the generics in other modules, these are meant to be imported to save you from the boilerplate
-
 T = TypeVar("T")
+out_T = TypeVar("out_T", covariant=True)
+in_T = TypeVar("in_T", contravariant=True)
 T_co = TypeVar("T_co", covariant=True)
 T_contra = TypeVar("T_contra", contravariant=True)
```
Comment (DetachHead) [inline basedtyping/__init__.py]: Why duplicate these? If we want to rename them we should add a note saying the old ones are deprecated

Comment (KotlinIsland) [reply basedtyping/__init__.py]: yeah, ig

File: .github/workflows/check.yaml [hunk 4]
```diff
@@ -3,14 +3,16 @@ name: check
...
 
 jobs:
   check:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", '3.12-dev']
+        include:
+          - usable-python-version: : ["3.8", "3.9", "3.10", "3.11", '3.12']
```
Comment (DetachHead) [inline .github/workflows/check.yaml]: how does this work

File: .github/workflows/check.yaml [hunk 5]
```diff
@@ -3,14 +3,16 @@ name: check
...
 
 jobs:
   check:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12-dev"]
+        include:
+          - usable-python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
Comment (DetachHead) [inline .github/workflows/check.yaml]: what this means

File: .vscode/extensions.json [hunk 6]
```diff
@@ -2,9 +2,7 @@
   "recommendations": [
     "ms-python.python",
     "ms-python.vscode-pylance",
-    "ms-python.pylint",
     "ms-python.black-formatter",
-    "ms-python.isort",
```
Comment (DetachHead) [inline .vscode/extensions.json:7]: add `charliermarsh.ruff`

File: .github/workflows/check.yaml (hunk 5 above)
Comment (KotlinIsland) [reply .github/workflows/check.yaml]: hmm, check again

**Review by DetachHead — APPROVED** [review]

**Pull request merged**